plugin: Transaction-based speculative reserve and logic unification

sharnoff opened the pull request:
Trying out this approach as a potential way to make implementing overcommit (#517) easier, to avoid dealing with rounding errors when responding to the autoscaler-agents.

I think this "speculative reserve" may be a reasonable change in isolation, and reducing the number of separate implementations of the same logic should also help with overcommit (because of how it affects every single resource calculation).

Not 100% sure we want to go this route, but it's my best guess so far.

---

Tested this locally and it seemed to be ok; already fixed a couple bugs based on suspicious logs. There might be more.

Omrigan reviewed (commented):
Partial review

Omrigan commented on pkg/plugin/trans.go:
```diff
@@ -64,81 +64,65 @@ func (s verdictSet) MarshalLogObject(enc zapcore.ObjectEncoder) error {
-
-	r.node.Reserved += r.pod.Reserved
-	r.node.Buffer += r.pod.Buffer
+	callback := func(oldState, newState resourceState[T]) string {
```
As discussed, perhaps all verdict formatting should be gathered at a single place, like a structure `verdictGenerator`, which will be used by `handleRequestedGeneric`.

sharnoff replied:
I like that idea -- I think it _may_ end up somewhat complex; wdyt about leaving it as a follow-up?

Omrigan commented on pkg/plugin/trans.go:
```diff
@@ -150,26 +134,130 @@ func (r resourceTransitioner[T]) handleReserve() (overbudget bool, verdict strin
 // Any permitted increases are required to be a multiple of factor.
 //
 // A pretty-formatted summary of the outcome is returned as the verdict, for logging.
```
It would be easier to understand, if those comments would more clearly reflect the difference between `handleRequested` and `handleReserve`.

sharnoff replied:
Fair enough - added some comments in bbfca613b2a4fa8b8d650625b4049296475c89f6, although not sure if there's more I should elaborate on.

sharnoff commented on pkg/plugin/trans.go:
```diff
@@ -117,6 +117,33 @@ func (r resourceTransitioner[T]) handleLastPermit(lastPermit T) (verdict string)
+		)
+	}
+
+	overbudget = r.node.Reserved > r.node.Total
```
`Buffer` is included in `Reserved`, so it should be ok as is

sharnoff commented on pkg/util/xact/xact.go:
```diff
@@ -0,0 +1,31 @@
+// New returns a new transaction object (called Xact) operating on the given pointer
+func New[T any](ptr *T) *Xact[T] {
+	return &Xact[T]{
+		tmp:  *ptr,
```
That's fair. Added a note about this in 0972c4ce6e1af352f4d6423cb99a97281dde3c65 -- the types I intended to use for this don't have pointers (and at least IMO, that's a reasonable restriction for now).

sharnoff commented:
Updates:

* Rebased onto main
* Add note about limitations w.r.t. pointers: 0972c4ce6e1af352f4d6423cb99a97281dde3c65
* Added comments about the difference between `handleRequested` vs `handleReserve`: bbfca613b2a4fa8b8d650625b4049296475c89f6
